build-common/cmake/PreDepends.cmake

```diff
@@ -0,0 +1,113 @@
+##############################################################################
+# PreDepends.cmake
+#
+# Macros to cope with the fact that cmake doesn't support the notion
+# of dependencies that must be built before any other target when
+# doing parallel builds.
+#
+# SYNOPSIS:
+#
+# Include this file in your top level CMakeLists.txt and call
+#
+#   PredependsInit()
+#
+# precisely once.
+#
+# Then, use
+
+#   ExternalDependency_Add(<NAME> BUILD_BYPRODUCTS <BYPRODUCTS> ARGS ...)
+# instead of
+#   ExternalProject_Add(<NAME> BUILD_BYPRODUCTS <BYPRODUCTS> ...)
+
+#   xadd_executable(<NAME> ...)
+# instead of
+#   add_executable(<NAME> ...)
+
+#   xadd_library(<NAME> ...)
+# instead of
+#   add_library(<NAME> ...)
+
+
+# This will ensure that the external dependencies get built before
+# any targets added with xadd_{executable,library}.
+
+
+# METHODS:
+# This module sets up a dummy target "__bc_predepends" and provides a
+# macro:
+#
+#   add_predepend(foo)
+#
+# This macro makes the __bc_predepends target depend on foo. It can
+# be used directly, e.g.,
+#
+#   ExternalProject_Add(foo BUILD_BYPRODUCTS path/to/libfoo.a ...)
+#   add_predepend(foo)
+#
+# but the typical usage is to call
+#   ExternalDependency_Add(foo BUILD_BYPRODUCTS path/to/libfoo.a ARGS ...)
+# instead.
+#
+# We provide another set of macros:
+#
+#   xadd_executable(bar bar.c)
+#   xadd_library(baz baz.c)
+#
+# These function exactly like the native add_executable/add_library
+# commands, but they also add a dependency on __bc_predepends
+# to the target being generated (bar and baz in this case).
+# This forces any registered predepends targets to be built first.
+# In the example at hand, we get the following dependency graph:
+#
+# bar -----------> __bc_predepends -> foo
+# libbaz.a ----`
+#
+# where a -> b is pronounced "a depends on b".
+#
+# NOTE: cmake 3.2+ is required to use the Ninja generator
+cmake_minimum_required(VERSION 2.8)
+
+set(PREDEPENDS_TARGET_NAME "__bc_predepends")
```

Comment:
It rhymes with "guild shaman", and also indicates the name of a repo that this might eventually get pushed backed to. This is me being paranoid about namespace pollution in cmake.